### ISSUE TITLE
Investigate masked token modeling

### ISSUE THREAD

**Issue by clessig** [ecmwf/WeatherGenerator#243]
### What happened?

Masking rate seems to be wrong

### What are the steps to reproduce the bug?

train with mtm

### Version

develop

### Platform (OS and architecture)

Linux

### Relevant log output

```shell

```

**Comment by kctezcan** [comment]
comment

**Comment by kctezcan** [comment]
I checked the generation of the masking rate with masking rate sampling. The variance of the distribution for sampling the masking rate seems to be too big. See the code snippet to see the samples masking rates (copy psated and adapted from the tokenizer_masking.py):

```python
import numpy as np
import time
rng = np.random.default_rng(int(time.time()))
mr = 0.6

N = 20
smrs = np.zeros(N)
for ix in range(10):
    smr = np.clip(np.abs(rng.normal(loc=mr, scale=1.0 / 2.5 * np.pi)), 0.0, 1.0)
    smrs[ix] = smr
    print(smr)
    
print(np.mean(smrs), np.std(smrs))
print(smrs)
``` 

This returns:
1.0
0.505522663620754
0.02652278505138639
0.06560116857201537
0.5625362002275125
0.3424484366862031
1.0
0.498775720205638
1.0
0.5197246024061427
0.2760565788384826 0.366675894370429
[1.         0.50552266 0.02652279 0.06560117 0.5625362  0.34244844
 1.         0.49877572 1.         0.5197246  0.         0.
 0.         0.         0.         0.         0.         0.
 0.         0.        ]

As you can see the sampled masking rate is changing too much. So we need to reduce the variance of the sampling is the hypothesis.

**Comment by clessig** [comment]
Yes, this is too large. Can you try with `scale=1.0 / (2.5 * np.pi)` ?

**Comment by kctezcan** [comment]
that gives:
1.0 / (2.5 * np.pi) = 0.1273

and the result looks better :):
used scale: 0.12732395447351627 | mean: 0.6067092453536115, std: 0.1428857589395425, max: 1.0, min: 0.151801067164627

Is that formula coming from somewhere?

**Comment by clessig** [comment]
The variance was tuned by hand, but I think the brackets got lost when I transferred it. We can explore other distributions but my priority (by a margin) would be on other masking strategies rather than tuning this sampling.

**Comment by clessig** [comment]
> -> This checks if the whole source is empty for the given time index. But maybe the error is more granular, lke with empty patches, right?

No, the empty target error only occurs when every patch is empty. But it was made more likely by the problem. But should be fixed anyway.

**Comment by kctezcan** [comment]
So can it be sth like this:
1. `np.array([s.target_empty() for s in streams_data]).all()` is `False` because there is data when you check in multi_stream_data_sampler.py line 363.
2. but this data is Nan and this is caught by `if torch.isnan(tc_tokens).any():` in line 584 in model.py with the error: "Skipping prediction for ERA5 because of 926328 NaN in tc_tokens."
3. and this if statement sets the preds_tokens to an empty tensor: `preds_tokens += [torch.tensor([], device=tc_tokens.device)]` in line 589 in model.py
4. So the check in step 1 is actually not catching the situation where data size is >0 but it is Nans?

The other question would then be why there are Nans in the era5 dataset at all...

**Comment by clessig** [comment]
> So can it be sth like this:
> 
> 1. `np.array([s.target_empty() for s in streams_data]).all()` is `False` because there is data when you check in multi_stream_data_sampler.py line 363.
> 2. but this data is Nan and this is caught by `if torch.isnan(tc_tokens).any():` in line 584 in model.py with the error: "Skipping prediction for ERA5 because of 926328 NaN in tc_tokens."

This should only happen if the prediction head/decoder network diverged.
